src/utils/sqs_consumer.ts

```diff
@@ -0,0 +1,89 @@
+import { SQS } from 'aws-sdk';
+
+import { ONE_SECOND_MS } from '../constants';
+import { logger } from '../logger';
+
+import { SqsClient } from './sqs_client';
+
+export class SqsConsumer {
+    private readonly _sqsClient: SqsClient;
+    private readonly _beforeHandle?: () => Promise<boolean>;
+    private readonly _handleMessage: (message: SQS.Types.Message) => Promise<any>;
+    private readonly _afterHandle?: () => Promise<void>;
+    private _isConsuming: boolean;
+
+    constructor(params: {
+        sqsClient: SqsClient;
+        beforeHandle?: () => Promise<boolean>;
+        handleMessage: (message: SQS.Types.Message) => Promise<any>;
+        afterHandle?: () => Promise<void>;
+    }) {
+        this._sqsClient = params.sqsClient;
+        this._beforeHandle = params.beforeHandle;
+        this._handleMessage = params.handleMessage;
+        this._afterHandle = params.afterHandle;
+        this._isConsuming = false;
+    }
+
+    public stop(): void {
+        this._isConsuming = false;
+    }
+
+    public async consumeAsync(): Promise<void> {
+        this._isConsuming = true;
+        while (this._isConsuming) {
+            await this.consumeOnceAsync();
+        }
+    }
+
+    public async consumeOnceAsync(): Promise<void> {
+        // Run the before hook
+        if (this._beforeHandle) {
+            let beforeCheck;
+            try {
+                beforeCheck = await this._beforeHandle();
+            } catch (e) {
+                logger.warn(e, 'Error encountered in the preHandle check');
+                return;
+            }
+
+            if (!beforeCheck) {
+                logger.warn('before validation failed');
```

Comment:
Do we want to sleep if the validation function failed (or returned `false`)? We know for a fact that this happens when the bot has no ETH or there is a pending transaction, so would it make sense to sleep for a reasonable amount of time like 1s?

Reply:
yeah, that's a good idea